clojure-mode.el

```diff
@@ -1824,6 +1836,114 @@ When BUT-LAST is passed the last expression is not threaded."
   (interactive "P")
   (clojure--thread-all "->> " but-last))
 
+;;; Cycling stuff
+
+(defcustom clojure-use-metadata-for-privacy nil
+  "If nil, `clojure-cycle-privacy' will use (defn- f []).
+If t, it will use (defn ^:private f [])."
+  :package-version '(clojure-mode . "5.5.0")
+  :safe #'booleanp
+  :type 'boolean)
+
+;;;###autoload
+(defun clojure-cycle-privacy ()
+  "Make public the current private def, or vice-versa.
+
+See: https://github.com/clojure-emacs/clj-refactor.el/wiki/cljr-cycle-privacy"
+  (interactive)
+  (save-excursion
+    (ignore-errors (forward-char 7))
+    (search-backward-regexp "\\((defn-? \\)\\|\\((def \\)")
+    (cond
+     ((and clojure-use-metadata-for-privacy
+           (looking-at "(defn ^:private"))
+      (forward-char 5)
+      (delete-char 10))
+     ((and (not clojure-use-metadata-for-privacy)
+           (looking-at "(defn-"))
+      (forward-char 5)
+      (delete-char 1))
+     ((and clojure-use-metadata-for-privacy
+           (looking-at "(defn"))
+      (forward-char 5)
+      (insert " ^:private"))
+     ((and (not clojure-use-metadata-for-privacy)
+           (looking-at "(defn"))
+      (forward-char 5)
+      (insert "-"))
+     ((looking-at "(def ^:private")
+      (forward-char 5)
+      (delete-char 10))
+     ((looking-at "(def ")
+      (forward-char 5)
+      (insert "^:private ")))))
+
+;;;###autoload
+(defun clojure-cycle-collection ()
+  "Convert the coll at (point) from (x) -> {x} -> [x] -> -> #{x} -> (x) recur.
+
+See: https://github.com/clojure-emacs/clj-refactor.el/wiki/cljr-cycle-coll"
+  (interactive)
+  (save-excursion
+    (while (and
+            (> (point) 1)
+            (not (eq (string-to-char "(") (char-after)))
+            (not (string= "#{" (buffer-substring (point) (+ 2 (point)))))
+            (not (eq (string-to-char "{") (char-after)))
+            (not (eq (string-to-char "[") (char-after))))
+      (backward-char))
+
+    (cond
+     ((eq (string-to-char "(") (char-after))
+      (insert "{" (substring (clojure-delete-and-extract-sexp) 1 -1) "}"))
+
+     ((eq (string-to-char "#") (char-after))
+      (delete-char 1)
+      (insert "(" (substring (clojure-delete-and-extract-sexp) 1 -1) ")"))
+
+     ((eq (string-to-char "{") (char-after))
+      (if (not (equal (string-to-char "#") (char-before)))
+          (insert "[" (substring (clojure-delete-and-extract-sexp) 1 -1) "]")
+        (backward-char)
+        (delete-char 1)
+        (insert "(" (substring (clojure-delete-and-extract-sexp) 1 -1) ")")))
+
+     ((eq (string-to-char "[") (char-after))
+      (insert "#{" (substring (clojure-delete-and-extract-sexp) 1 -1) "}"))
+
+     ((equal 1 (point))
+      (user-error "Beginning of file reached, collection is not found")))))
+
+(defun clojure--goto-if ()
+  (when (in-string-p)
+    (while (or (not (looking-at "("))
+               (in-string-p))
+      (backward-char)))
+  (while (not (looking-at "\\((if \\)\\|\\((if-not \\)"))
```

Comment:
FTR, I like this method of going backward-up-list until we find what we're looking for (instead of doing search-backward). And I think this is what we should be using on all the other commands too (including the thread/unwind commands). But that's something for another time.